File: web/routes/discovery.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Request

from web.database import get_db
from web.models import DiscoveryTriggerRequest
from web.services.discovery_service import DiscoveryService

router = APIRouter()


def _discovery_service(request: Request, db=Depends(get_db)) -> DiscoveryService:
    return DiscoveryService(db, request.app.state.obs_client)


@router.post("/trigger")
async def trigger_discovery(
    body: DiscoveryTriggerRequest, service: DiscoveryService = Depends(_discovery_service),
):
    return await service.trigger(body.full_catalog, body.force)


@router.get("/history")
async def discovery_history(service: DiscoveryService = Depends(_discovery_service)):
    return await service.get_history()


@router.get("/new")
async def new_posts(service: DiscoveryService = Depends(_discovery_service)):
    return await service.get_new_posts()

File: frontend/src/api/discovery.ts
import type { DiscoveryRunResponse, PostSummary } from '../types'
import { api } from './client'

export const discoveryApi = {
  trigger: (fullCatalog = false, force = false) =>
    api.post<DiscoveryRunResponse>('/api/discovery/trigger', { full_catalog: fullCatalog, force }),
  history: () => api.get<DiscoveryRunResponse[]>('/api/discovery/history'),
  newPosts: () => api.get<PostSummary[]>('/api/discovery/new'),
}
